### PR TITLE
Fix: dont double-fetch when adding a pin

### PR DESCRIPTION
**Pull request by dirkmc**
Also error-out early if trying to direct-pin a key that is already recursively pinned

See #6428

### CODE REVIEW

File: pin/pin.go [hunk 1]
```diff
@@ -213,11 +213,6 @@ func NewPinner(dstore ds.Datastore, serv, internal ipld.DAGService) Pinner {
 func (p *pinner) Pin(ctx context.Context, node ipld.Node, recurse bool) error {
 	p.lock.Lock()
 	defer p.lock.Unlock()
-	err := p.dserv.Add(ctx, node)
```
Comment (Stebalien) [inline pin/pin.go]: I believe we still need to _add_ this node, we can probably just do it outside the lock.

Comment (dirkmc) [reply pin/pin.go]: I basically reverted [this commit](https://github.com/ipfs/go-ipfs/commit/953b610470044bd332522b346e4881b8957c2050#diff-10c79301e027d970f7cba49f4a853d86), ie I'm assuming that we have already added before `Pin()` is called.
Is that incorrect?

Comment (Stebalien) [reply pin/pin.go]: I really don't know. The safe solution would be to change the API. Let's pause this for a few days while I try to pull in a CoreAPI patch, then we can look into making `Pin` take a CID.

Comment (gammazero) [reply pin/pin.go:216]: This is done outside the lock now:
https://github.com/ipfs/go-ipfs-pinner/blob/master/dspinner/pin.go#L177

File: pin/pin.go [hunk 2]
```diff
@@ -246,6 +241,10 @@ func (p *pinner) Pin(ctx context.Context, node ipld.Node, recurse bool) error {
 
 		p.recursePin.Add(c)
 	} else {
+		if p.recursePin.Has(c) {
+			return fmt.Errorf("%s already pinned recursively", c.String())
+		}
+
```
Comment (gammazero) [inline pin/pin.go:247]: This is also done in the current pinner:
https://github.com/ipfs/go-ipfs-pinner/blob/master/dspinner/pin.go#L246-L247

**Comment by gammazero** [comment]
Closing: These changes are already in the current pinner.

**Pull request closed without merging**